### ISSUE TITLE
perf regression in 11e5af0c66cea714c

### ISSUE THREAD

**Issue by JustinAzoff** [google/gopacket#372]
After we fixed those memory allocations today, I upgraded gopacket on our box.. which fixed the memory allocations, but the cpu usage went through the roof.  I was able to bisect things down to this commit.

```
commit 11e5af0c66cea714cf0d2cc4c8f1ed61d0dc6a2f
Date:   Fri Nov 18 08:12:26 2016 -0800

    pcap: make stoppable when no packets to sniff
```

I'm not sure why specifically it impacts things so much on live devices, but that's definitely the cause.

sniff.go is the same test program from #367

Before, 2300ms and 5million instructions or so

```
~ $ (cd go/src/github.com/google/gopacket/;git checkout a12eb5bddccf95ff01d94445993b768547a33a62)
Previous HEAD position was 11e5af0... pcap: make stoppable when no packets to sniff
HEAD is now at a12eb5b... [afpacket] Avoid sending empty block
~ $ go build sniff.go ;perf stat ./sniff snf1
2017/10/10 20:59:52 Alloc = 18041 TotalAlloc = 19465 Sys = 23238 NumGC = 3
2017/10/10 20:59:52 Alloc = 21219 TotalAlloc = 22643 Sys = 26730 NumGC = 3
2017/10/10 20:59:53 Alloc = 20098 TotalAlloc = 25774 Sys = 29226 NumGC = 4
2017/10/10 20:59:54 Alloc = 23229 TotalAlloc = 28905 Sys = 30286 NumGC = 4
2017/10/10 20:59:54 Alloc = 26356 TotalAlloc = 32032 Sys = 33518 NumGC = 4
2017/10/10 20:59:55 Alloc = 29483 TotalAlloc = 35159 Sys = 36754 NumGC = 4
2017/10/10 20:59:55 Alloc = 32612 TotalAlloc = 38287 Sys = 39990 NumGC = 4
2017/10/10 20:59:56 Alloc = 19248 TotalAlloc = 41418 Sys = 42230 NumGC = 5
2017/10/10 20:59:56 Alloc = 22377 TotalAlloc = 44547 Sys = 42294 NumGC = 5
2017/10/10 20:59:57 Alloc = 25502 TotalAlloc = 47672 Sys = 42358 NumGC = 5

 Performance counter stats for './sniff snf1':

       2295.810533      task-clock (msec)         #    0.343 CPUs utilized
     5,072,239,542      instructions              #    0.79  insn per cycle
       6.689448392 seconds time elapsed

```

After, 6000ms and 23million instructions

```
~ $ (cd go/src/github.com/google/gopacket/;git checkout 11e5af0c66cea714cf0d2cc4c8f1ed61d0dc6a2f)
Previous HEAD position was a12eb5b... [afpacket] Avoid sending empty block
HEAD is now at 11e5af0... pcap: make stoppable when no packets to sniff
~ $ go build sniff.go ;perf stat ./sniff snf1
2017/10/10 20:58:58 Alloc = 18058 TotalAlloc = 19463 Sys = 24038 NumGC = 3
2017/10/10 20:58:58 Alloc = 21740 TotalAlloc = 26931 Sys = 27402 NumGC = 4
2017/10/10 20:58:58 Alloc = 18903 TotalAlloc = 40612 Sys = 41974 NumGC = 5
2017/10/10 20:58:59 Alloc = 26964 TotalAlloc = 65204 Sys = 42422 NumGC = 6
2017/10/10 20:58:59 Alloc = 18157 TotalAlloc = 89395 Sys = 42678 NumGC = 8
2017/10/10 20:59:00 Alloc = 17643 TotalAlloc = 122009 Sys = 42934 NumGC = 10
2017/10/10 20:59:00 Alloc = 25211 TotalAlloc = 146214 Sys = 42934 NumGC = 11
2017/10/10 20:59:01 Alloc = 22383 TotalAlloc = 176510 Sys = 42934 NumGC = 13
2017/10/10 20:59:01 Alloc = 24423 TotalAlloc = 195019 Sys = 42934 NumGC = 14
2017/10/10 20:59:02 Alloc = 21233 TotalAlloc = 225031 Sys = 43994 NumGC = 16

 Performance counter stats for './sniff snf1':

       5986.143775      task-clock (msec)         #    1.088 CPUs utilized
    22,757,729,356      instructions              #    1.31  insn per cycle
       5.504411887 seconds time elapsed
```

and for completeness.. master

```
~ $ (cd go/src/github.com/google/gopacket/;git checkout master)
Previous HEAD position was 11e5af0... pcap: make stoppable when no packets to sniff
Switched to branch 'master'
~ $ go build sniff.go ;perf stat ./sniff snf1
2017/10/10 20:58:03 Alloc = 18087 TotalAlloc = 19480 Sys = 24294 NumGC = 3
2017/10/10 20:58:05 Alloc = 18163 TotalAlloc = 19556 Sys = 24294 NumGC = 3
2017/10/10 20:58:06 Alloc = 18165 TotalAlloc = 19558 Sys = 24294 NumGC = 3
2017/10/10 20:58:07 Alloc = 18171 TotalAlloc = 19564 Sys = 24294 NumGC = 3
2017/10/10 20:58:08 Alloc = 18187 TotalAlloc = 19580 Sys = 24294 NumGC = 3
2017/10/10 20:58:08 Alloc = 18187 TotalAlloc = 19580 Sys = 24294 NumGC = 3
2017/10/10 20:58:08 Alloc = 18187 TotalAlloc = 19580 Sys = 24294 NumGC = 3
2017/10/10 20:58:09 Alloc = 18187 TotalAlloc = 19580 Sys = 24294 NumGC = 3
2017/10/10 20:58:09 Alloc = 18187 TotalAlloc = 19580 Sys = 24294 NumGC = 3
2017/10/10 20:58:10 Alloc = 18187 TotalAlloc = 19580 Sys = 24294 NumGC = 3

 Performance counter stats for './sniff snf1':

       7567.871153      task-clock (msec)         #    0.991 CPUs utilized
    30,138,120,652      instructions              #    1.40  insn per cycle

       7.635663550 seconds time elapsed
```

I did start working on a myricom packet source.. I only implemented the bare minimum, but so far it works, and the overhead is pretty low, as a comparison:

```
       1143.013177      task-clock (msec)         #    0.228 CPUs utilized
     3,019,991,539      instructions              #    0.96  insn per cycle
 
       5.006608108 seconds time elapsed
```

Should get even lower if I can figure out how to use snf_ring_recv_many.

Anyway, it looks like just turning nonblocking off and leaving everything else the same fixes things:

```
-       if v := C.pcap_setnonblock(p.cptr, 1, buf); v == -1 {
+       if v := C.pcap_setnonblock(p.cptr, 0, buf); v == -1 {
```

So if there was a way to have an option that toggles that, that wouldn't require refactoring everything again or having two versions of the packet source

**Comment by JustinAzoff** [comment]
perf graphs showing the difference:

<img width="821" alt="screen shot 2017-10-10 at 10 32 28 pm" src="https://user-images.githubusercontent.com/65812/31419666-ff291318-ae0a-11e7-9ad6-399467d73846.png">

top graph is allocations that the earlier fixes knocked to almost zero.. bottom graph is total cpu, which maxed out with the new code.

After toggling nonblocking back on, everything is back to where it was before.

**Comment by jcrussell** [comment]
See #339.

**Comment by JustinAzoff** [comment]
Fixed as part of #339 !
